### PR TITLE
feat(common/models): add Trie string-encoding + decoding methods 💾

### PR DESCRIPTION
**Pull request by jahorton**
This PR aims to help address #10336, though not sufficient for a full fix.

So far, this PR establishes methods useful to "compress" / encode our lexical model Tries into a notably more compact format and to "decompress" / decode them from that format, one piece at a time.

It does _not_ currently:
- link these methods into our existing Trie definitions
- use them at run-time
- use them when compiling lexical models

So, there's obviously more work that would be needed, but it's a solid start in the right direction.

----

Possibly worth considering:  do we build in some reserved ranges now in case we need future extensibility?  See https://github.com/keymanapp/keyman/pull/11088#issuecomment-2276991362.

@keymanapp-test-bot skip

### CODE REVIEW

File: common/models/templates/src/tries/compression.ts [hunk 1]
```diff
@@ -0,0 +1,273 @@
+import { Entry, InternalNode, Leaf, Node } from "../trie-model.js";
+
+// const SINGLE_CHAR_RANGE = Math.pow(2, 16) - 64;
+// const ENCODED_NUM_BASE = 64;
+
+// Offsetting by even just 0x0020 avoids control-code chars + avoids VS Code not liking the encoding.
+const ENCODED_NUM_BASE = 0x0000;
+const SINGLE_CHAR_RANGE = Math.pow(2, 16) - ENCODED_NUM_BASE;
```
Comment (jahorton) [inline common/models/templates/src/tries/compression.ts:8]: A notable differentiation from the pseudo-spec established in #10336.  It does restrict the data ranges slightly, but it makes a _big_, positive difference in the encoding and how IDEs interpret the resulting encoded Trie data when written to files.

Note:  the unit tests established later currently do not adjust for alternate `ENCODED_NUM_BASE` values.  This shouldn't be too tricky to establish for reasonable value selections, though.

File: common/models/templates/test/fixtures/tries/english-1000.json
Comment (jahorton) [inline common/models/templates/test/fixtures/tries/english-1000.json:1]: A fixed version of the fixture, utilizing #11074.  My issues using this fixture during development of this PR were the cause of #11073's discovery.

File: common/models/templates/test/trie-compression.js [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+    const encodedSerializableString = `${
+      compressedTrie
+        .replace(/\\/g, '\\\\') // preservation - escape char as literal
+        .replace(/`/g, '\\`') // escape the primary quote
+        .replace(/\n/g, '\\n') // prevent CRLF shenanigans
+        .replace(/\r/g, '\\r')
+    }`;
+
+    const compression = {
+      // Achieves FAR better compression than JSON.stringify, which \u-escapes most chars.
```
Comment (jahorton) [inline common/models/templates/test/trie-compression.js:235]: Unless using `ENCODED_NUM_BASE=0x0020` or similar.  `JSON.stringify` likes to `\u`-escape control characters, which leads to notable string-bloat when the control characters are utilized - they tend to represent values that appear with high frequency in the encoding.

Comment (jahorton) [reply common/models/templates/test/trie-compression.js:235]:
```
> JSON.stringify(String.fromCharCode(1))
'"\\u0001"'
```

With `ENCODED_NUM_BASE=0`...

- note that most leaf notes will have low, single-digit `.entries` counts, all of which would be represented by control codes and thus subject to `\u`-escaping.  The word lengths will usually be notably less than 32 chars and thus would also subject to the same effects... leading to most encoded entries having length less than 32 chars.

- also, most near-leaf internal nodes will have but a few legal `values` leading to child nodes, once again using control codes for their representation.

JSON.stringify use is much prettier and straightforward with `ENCODED_NUM_BASE=0x0020` or above, as this bypasses the control-code range with one exception: 0x007f (DEL).

**Comment by jahorton** [comment]
One potential issue that comes to mind:  we may wish to add fields for entries and nodes to allow for future extensibility, should that become needed.

`Entry` currently uses the 'standard' header length (used for entries, leaves, and internal nodes), which uses two UTF-16 code units to encode a length.  I'd be extremely shocked to have a single `Entry` actually need over 64,000 characters... so we _could_ just repurpose (and reserve) one of those chars for a secondary use, to signal an alternate layout.  (The two-char width was originally specified due to top-level internal nodes needing lots of space to include their children.)  An `Entry` can never exist in places where a `Node` (leaf, internal) is expected, so using a different width for the `Entry` section width ought be somewhat safe.
- Since no `Entry` should even approach capping a 16-bit integer (65536) length... if we're willing to commit to a stricter range, we could reserve a couple of bits instead on the encoded length's char while still shifting to single-char length for Entries.

As for leaves and internal nodes, both track a 'count' within a single char.  That said, something's gone wrong if we need 32,000 direct entries or children... let alone a quarter of that.  We could reserve a few extra bits on that char for future extensions to the two types without increasing the number of bytes needed to store either.

**Comment by mcdurdin** [comment]
> do we build in some reserved ranges now in case we need future extensibility?

No, I think we just ensure that the data structure is versioned. Then we can change format if we need to by bumping the version.

**Review by mcdurdin — APPROVED** [review]

**Comment by jahorton** [comment]
> > do we build in some reserved ranges now in case we need future extensibility?
> 
> No, I think we just ensure that the data structure is versioned. Then we can change format if we need to by bumping the version.

This would mean that the new versions are assumed 'unreadable' by older versions.  A bit of work would avoid that issue, letting old-version compatible code to read the parts of the data structure it recognizes while ignoring the parts it hasn't yet seen.

We're already talking about a breaking-change format moving to 18.0, where 17.0 and before won't be able to load models with a Trie-encoding format based on these changes.  I would prefer to minimize the risk of causing that again.

That said, we can always leave it as a future work item, linking my thoughts in that direction above as a note for implementation before 18.0 release.

**Comment by mcdurdin** [comment]
> This would mean that future new 'versions' are assumed 'unreadable' by older versions. A bit of work would avoid that issue, letting old-version compatible code to read the parts of the data structure it recognizes while ignoring the parts it hasn't yet seen.
>
> We're already talking about a breaking-change format moving to 18.0, where 17.0 and before won't be able to load models with a Trie-encoding format based on these changes. I would prefer to minimize the risk of causing that again.

This is premature design. Do you have a specific use case that you are designing for where you can demonstrate that this will be sufficient to support that feature? If not, then reserving the space may not end up being adequate anyway, and we may still need a newer version of Keyman to support the new functionality -- and we've done this premature design for nothing.

We had to play these games in the past when upgrading was more difficult. The less we have to do it going forward, the better.

Keyman is becoming increasingly evergreen, and reserving space for an unknown future use is not the right thing to do -- either in terms of file formats, or in terms of APIs, or in internal design. Design for use now. Refactor when needed. Have good version management -- and that is sufficient.

**Comment by jahorton** [comment]
> > This would mean that future new 'versions' are assumed 'unreadable' by older versions. A bit of work would avoid that issue, letting old-version compatible code to read the parts of the data structure it recognizes while ignoring the parts it hasn't yet seen.
> > We're already talking about a breaking-change format moving to 18.0, where 17.0 and before won't be able to load models with a Trie-encoding format based on these changes. I would prefer to minimize the risk of causing that again.
> 
> This is premature design. Do you have a specific use case that you are designing for where you can demonstrate that this will be sufficient to support that feature? If not, then reserving the space may not end up being adequate anyway, and we may still need a newer version of Keyman to support the new functionality -- and we've done this premature design for nothing.
> 
> We had to play these games in the past when upgrading was more difficult. The less we have to do it going forward, the better.
> 
> Keyman is becoming increasingly evergreen, and reserving space for an unknown future use is not the right thing to do -- either in terms of file formats, or in terms of APIs, or in internal design. Design for use now. Refactor when needed. Have good version management -- and that is sufficient.

**Example 1: localized emoji-oriented models**  

Noting the structure of the package I referenced for #905, they actually include use of GitHub and Slack style emoji code input:  `:emoji-name-here:`.  That works well... for **English**.  These would probably be localized for different languages, and likely should be possible to look up based on their localized name.  This is _also_ separate from their actual encoded emoji character, the primary associated output.

To be clear, that package also supports keywords associated with the emojis - the prefix of any associated keyword will suggest the associated emoji.  We'd likely have multiple entries per emoji - one per keyword for the emoji + its full-length standard tag - all of which should also provide the associated emoji... or the English tag for that emoji.  Either version - the emoji codepoint or the English tag - is not closely associated with the key or the localized emoji tag - and will likely motivate an extra field.

If desired, we should be able to create Crowdin localization files based on the backing data's associated JSON structure and programmatically construct them if using the "English tag" variant.  (That "English tag" would serve as the string's localization key.)  This would allow progressive localization of emojis as a group effort by each language community, in theory, as a potential approach.

**Example 2: Agglutinative models**

In agglutinative models, we'll likely want to store affixes in tries for lookup.  Granted, it's unclear if we'll want one Trie per affix type or all affixes in the same Trie... but if it's the latter, we may wish to 'tag' the affix's type with its entry.

Even if not _that_, I imagine word roots will have some sort of list of legal associated affixes - and that would likely need to be tagged onto each root's Trie entry.

**Comment by jahorton** [comment]
I guess I'm probably conflating reuse of the same code paths, with extensions for specific versions, with a desire for backward compatibility.  Could always have a "new version" later that uses the same functions where they apply, then adds extra sections to work with the new data.

What I'm really hoping for with that:  to avoid the need for _major variations_ in the encoding patterns as we make more specific versions in the future.  Not so much for the 18.0 release to actually _use_ extended data from Trie versions that don't yet exist.

**Comment by mcdurdin** [comment]
> You say this, but we've taken great pains to maintain backward compatibility for keyboards. Last I checked, we still actively support KMW 1.0 and 2.0 keyboards! Why are models being treated differently?

That's backward compatibility, not forward compatibility

File: common/models/templates/test/test-trie-compression.js [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+  describe('`Entry`s', () => {
+    it('compresses properly', () => {
+      assert.equal(compressEntry(TEST_DATA.ENTRIES.four.original), TEST_DATA.ENTRIES.four.compressed);
+    });
+  });
+
+  describe('Leaf nodes', () => {
+    it('compresses (mocked Entry)', () => {
+      // Should not attempt to recompress the mock-compressed entry.
+      assert.equal(compressNode(TEST_DATA.LEAVES.four.decompressed), TEST_DATA.LEAVES.four.compressed);
```
Comment (ermshiperete) [inline common/models/templates/test/test-trie-compression.js:139]: I do not see a difference between `TEST_DATA.LEAVES.four.decompressed` and `TEST_DATA.LEAVES.four.original`, so the comment doesn't make sense to me and I don't see the difference between the two tests here.

File: common/models/templates/test/test-trie-compression.js [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+      smpWordlist.forEach((tuple) => builder.addEntry(tuple[0], tuple[1]));
+
+      assert.doesNotThrow(() => builder.compress());
+    });
+
+    it.skip('properly round-trips a Trie through compression and decompression', () => {
+      // Requires full Trie integration, which isn't included yet.
+    });
+  });
+});
```
Comment (ermshiperete) [inline common/models/templates/test/test-trie-compression.js:260]: Do we need a test for decompressing `\ufdd0`?

**Review by ermshiperete — APPROVED** [review]

File: common/models/templates/src/trie-compression.ts [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+
+  const totalArr = [compressNumber(valueCntAndType)].concat(values).concat(compressedChildren);
+  return totalArr.join('');
+}
+
+function decompressInternal(str: string, baseIndex: number): Omit<InternalNode, 'children'> & { children: {[char: string]: string} } {
+  const weight = decompressNumber(str, baseIndex + NODE_SIZE_WIDTH, baseIndex + NODE_SIZE_WIDTH + WEIGHT_WIDTH);
+
+  // Assumes string-subsection size check has passed.
+  const childCnt = decompressNumber(str, baseIndex + NODE_TYPE_INDEX, baseIndex + NODE_TYPE_INDEX + 1);
```
[the remark this answers is not in the document]
Comment (jahorton) [reply common/models/templates/src/trie-compression.ts:231]: Line 195 for internal nodes.

```ts
const valueCntAndType = values.length;
```

File: common/models/templates/src/trie-compression.ts [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+  }
+
+  if(num) {
+    throw new Error(`Could not properly compress ${num} within specified char width of ${width}`);
+  }
+
+  return compressed;
+}
+
+const ENTRY_HEADER_WIDTH = NODE_SIZE_WIDTH + WEIGHT_WIDTH;
```
[the remark this answers is not in the document]
Comment (jahorton) [reply common/models/templates/src/trie-compression.ts:58]: Doc-comments have been added.

**Pull request merged**